feat: serve bundled functions in ntl dev

eduardoboucas opened the pull request:
**- Summary**

This PR uses `zip-it-and-ship-it` to bundle functions before serving them with `netlify dev`.

**- Test plan**

⏳ 

**- A picture of a cute animal (not mandatory but encouraged)**

![cute-lion-cub](https://user-images.githubusercontent.com/4162329/113895342-cc2dca00-97c0-11eb-82eb-da6c2f7b7927.jpg)

eduardoboucas commented on src/function-builder-detectors/zisi.js:
```diff
@@ -0,0 +1,67 @@
+  } catch (error) {
+    log(`${NETLIFYDEVERR} Could not create directory: ${targetDirectory}`)
+
+    process.exit(1)
```
Not sure whether to `exit` or let the error bubble up?

erezrokah replied:
To exit with an error we use `oclif` `error` utility, see https://github.com/netlify/cli/blob/f881b668de15bf5a9ec6988e5ce543caa7e22df8/src/commands/functions/create.js#L214

Regardless, I think this can be a warning as `netlify dev` will be partially useful?

eduardoboucas replied:
We won't be able to serve the functions if we can't write to the directory, though.

erezrokah replied:
We currently warn on function builder errors, so I think we can remove the `try/catch` here and let the CLI handle it:
https://github.com/netlify/cli/blob/f5b26d2295f46975194fc4c984e7695727b02029/src/utils/serve-functions.js#L409

eduardoboucas replied:
Changed in https://github.com/netlify/cli/pull/2094/commits/78f4ec0c541d8d21d7c3f2d42596c08aff87ef95 to use the `errorExit` property already being passed to the functions server. What do you think?

eduardoboucas replied:
Oh, I think we commented at the same time! 😅 

Just checking that you've seen https://github.com/netlify/cli/commit/78f4ec0c541d8d21d7c3f2d42596c08aff87ef95.

eduardoboucas commented on src/utils/serve-functions.js:
```diff
@@ -458,12 +461,24 @@ const startServer = async ({ server, settings, log, errorExit }) => {
   if (settings.functions) {
-    await setupFunctionsBuilder({ site, log, warn })
-    const server = await getFunctionsServer({ dir: settings.functions, siteUrl, capabilities, warn })
+    const { target: functionsDirectory } = await setupFunctionsBuilder({
```
The functions builder API now has the ability to return a `target` property, which defines the new functions directory.

erezrokah commented on src/function-builder-detectors/zisi.js:
```diff
@@ -0,0 +1,65 @@
+  return targetDirectory
+}
+
+module.exports = async function handler({ config, errorExit, functionsDirectory: sourceDirectory }) {
```
At the moment we evaluate detecters based on the order of `readdirSync`:
https://github.com/netlify/cli/blob/f5b26d2295f46975194fc4c984e7695727b02029/src/utils/detect-functions-builder.js#L6

I don't think it is guaranteed to have a consistent order.

1. Should we sort the files by name?
2. Should we add a condition to not use `zip-it-and-ship-it` builder when `netlify-lambda` is used?
3. Should we add a condition to use `zip-it-and-ship-it` only when someone is using `esbuild` (I think that would minimize the risk for the PR and we can later on make this behavior the default).

eduardoboucas replied:
Great points.

I have done 1 and 3 in https://github.com/netlify/cli/pull/2094/commits/1b2278b755c89a57965d5f7dc2d730d366079183.

I don't think we need 2, because `netlify-lambda` will be evaluated before `zisi`, due to the alphabetical order, and if it returns a truthy value, we stop looking for the next handler (https://github.com/netlify/cli/blob/feat/serve-bundled-functions/src/utils/detect-functions-builder.js#L18).

erezrokah commented on src/utils/serve-functions.js:
```diff
@@ -417,32 +425,43 @@ const getBuildFunction = ({ functionBuilder, log }) =>
-      leading: true,
-      trailing: true,
-    })
+  if (!functionBuilder) {
```
I like this early return!

erezrokah reviewed (changes requested):
Hi @eduardoboucas, this looks great pending one code comment.

What do you think about adding a test case for this?
It will be nice to test es6 modules and TypeScript functions as part of the CLI tests.

eduardoboucas commented:
> Hi @eduardoboucas, this looks great pending one code comment.
> 
> What do you think about adding a test case for this?
> It will be nice to test es6 modules and TypeScript functions as part of the CLI tests.

Thanks @erezrokah! I've addressed the comment and also added tests in https://github.com/netlify/cli/pull/2094/commits/88f1f47bf5e68589a8cecb626a464e89b1732884.

Let me know what you think!

erezrokah reviewed (approved):
🚀

eduardoboucas commented on src/utils/serve-functions.js:
```diff
@@ -417,32 +425,43 @@ const getBuildFunction = ({ functionBuilder, log }) =>
-  }
+  log(`${NETLIFYDEVLOG} Function builder ${chalk.yellow(functionBuilder.builderName)} detected${npmScriptString}.`)
+  warn(
+    `${NETLIFYDEVWARN} This is a beta feature, please give us feedback on how to improve at https://github.com/netlify/cli/`,
```
@erezrokah Do we still want to display this message for people using the ZISI function builder?

erezrokah replied:
Great catch - I forgot about it. Let's remove it all together

eduardoboucas replied:
Done in https://github.com/netlify/cli/pull/2094/commits/4ad71a4960407fc2397ac485c07acb331245b490.
